### ISSUE TITLE
Spatial translate (local space) vs translate_object_local are the same under the hood and not consistent with rotate/rotate_object_local naming convention.

### ISSUE THREAD

**Issue by viksl** [godotengine/godot#49370]
**Godot version:**
3.3.2

**OS/device including version:**
Windows 10

**Issue description:**

Spatial has three translate methods and three rotate methods:
1. global_translate | global_rotate
2. translate | rotate
3. translate_object_local | rotate_object_local

I'll start with rotation:
1. Rotates around global basis
2. Rotates around parent basis (local space)
3. Rotates around it's own basis (object local space)

However translation works like this:
1. Translates relative to global
2. Translates relative to object local space
3. Translates relative to object local space (but takes into account scale I think)

I'm not sure if this is intended, I was discussing it with people on discord who were checking the source code and also inputting different values to see how the 2. and 3. translation work and they seem to give the same results (as long as the scale is the same I think, I'm not sure about the scale sorry).

I'd expect considering the naming that the 2. and 3. translation would be similar to rotations in which space each works.

I apologize in case this is intended and not a bug.
(Note: Yes, I do know that Transform has rotated and translated)

I was brought to this by an article in the manual which uses translate_object_local to showcase strafing which from the text it seems it implies the behaviour should be as I expected above (same as the rotations).
https://docs.godotengine.org/en/stable/tutorials/3d/using_transforms.html

**Comment by aaronfranke** [comment]
Note that this functionality already exists, even if the methods are a bit confusing. If you want to translate relative to the parent, you can use `transform.origin += your_vector`.

**Comment by viksl** [comment]
> 
> 
> Note that this functionality already exists, even if the methods are a bit confusing. If you want to translate relative to the parent, you can use `transform.origin += your_vector` or `translation += your_vector`.

I know there are to achieve this but thanks for clarifying this in case someone else drops by.

My point here is more that if you open a manual and start checking how stuff works the way the methods currently are don't make sense for anyone who comes to godot, if it wasn't for the rotation methods I'd say well just 2 methods that do the same thing basically but since the naming is obviously made the same it just kind of doesn't make sense not to have the same pattern I guess?

I gave it several searches online, discussions on godot discord with plenty of people over several days and as far as I can tell everyone is just confused even people who were there working with godot and c++.

I don't know maybe there's a reason to keep this convention but for clarity and ease of use of anyone who comes to this (and that's going to be even more people with godot 4.0 getting closer and closer) I think it would be worth to make this work as suggested or change the docs and the names since right now I think it safe to say that after reading docs and the method names nobody would expect the results they give.

I hope this clarifies my point better.

**Comment by viksl** [comment]
@Calinou Please, could we bump this up from 3.x to 4.0 or so?
I understand that in 3.x it was a discussion since it was a breaking change but now with so many breaking changes in 4.0 maybe we could have the Translate fixed? :)

**Comment by YuriSizov** [comment]
The window for breaking changes in 4.0 has closed. We can make changes in 4.x only if they come with a clear migration path (preferably automated).

**Comment by aaronfranke** [comment]
Considering the current vs desired behavior, this can't be fixed without breaking compatibility, so it needs to wait until Godot 5.0.

However, this is a minor issue, you can perform the desired parent-relative translation as I mentioned above, so there is no missing functionality, just a confusing API.

**Comment by viksl** [comment]
Then could we make changes to the docs with a note in Translate it works in a different space perhaps, you know for now? I checked some other parts in docs and some misunderstandings there with this were fixed to reflect the behaviour. :-)

**Comment by YuriSizov** [comment]
Yes, adding clarity to the documentation is welcome in the meantime. Feel free to open a PR if you know how to make things clearer. (For the class reference, a PR is made against this repo, for the tutorials in online docs — in godot-docs).

**Comment by MajorMcDoom** [comment]
If we're going to make the documentation clearer, it should mention that this is a known bug. I'm afraid that once the documentation is clarified to essentially say that `translate` and `translate_object_local` do exactly the same thing, that this will cease to be considered a bug and we will lose context, adding further confusion down the line. That said, I've never seen the class reference mention known bugs. How should this be handled?

**Comment by kleonc** [comment]
> However translation works like this:
>     1. Translates relative to global
>     2. Translates relative to object local space
>     3. Translates relative to object local space (but takes into account scale I think)

To clarify, both `Spatial/Node3D::translate` (2) and `Spatial/Node3D::translate_object_local` (3) do exactly the same thing, there's no discrepancy in the behavior between them, both translate in the object local space.

---


I think this is a bug as clear as it can be, and should be fixed as such. To me it makes no sense to keep both methods doing the same thing "to not break compatibility". It's not like there is a single method which has mismatching naming/behavior (in such case it would make sense to keep it as is and only clarify the docs). There are already 2 different methods (`Node3D::translate` and `Node3D::translate_object_local`) named somehow according to their expected behavior, but one of them is unintentionally implemented to do the same thing as the other one.

Some recap:
- `Spatial::translate` was added in a93e33f5c8afc8ce2e9c1674718ff103bfae49b0, it was originally translating in the parent space (consistent with other added methods: `Spatial::rotate`, `Spatial::rotate_{x|y|z}`, `Spatial::scale`).
- `Spatial::translate` was changed in 9acab32daaea38e09c4d74d5f0340479d3dd41bd, seemingly for no reason, to use `Transform::translate` instead of `.origin += offset` ([here](https://github.com/godotengine/godot/commit/9acab32daaea38e09c4d74d5f0340479d3dd41bd#diff-4056cc1b21377c64b981832df445359279e6f3c8497fad5f310edf00c63490eaL629-R629)); this change made it translating in the local space instead of the parent one (that's how `Transform::translate` worked back then). I'm guessing this was unintentional change of the behavior. This introduced inconsistency with `Spatial::rotate`, `Spatial::rotate_{x|y|z}`, `Spatial::scale`.
- `Spatial::translate_object_local` was added in #15083, it was translating in the local space as intended. But looking at its implementation, and given that `Spatial::translate` was left unchanged, it seems like the author misinterpreted how `Transform::translate` worked back then (otherwise the implementation would be simpler and `Spatial::translate` would likely be changed).
- The rest are only renamings (including #57698), the implementation/behavior of both methods remained unchanged.

---

BTW if e.g. `Node3D::to_global` and `Node3D::to_local` were both doing the same thing I doubt anyone would think about keeping such buggy behavior unchanged for compat reasons. I'd say such bugs are meant to be fixed, not kept for compat... 🤔
